File: content/exercises/intro-r/data-frames/df-subset/checker.R
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 2) {
  stop("Usage: checker.R <submission_path> <output_path>")
}

submission_path <- args[[1]]
output_path <- args[[2]]

test_results <- list()
feedback <- c()
status <- "failed"

record_test <- function(name, passed, message) {
  test_results[[length(test_results) + 1]] <<- list(
    name = name,
    passed = passed,
    message = message
  )
}

env <- new.env(parent = baseenv())
execution_ok <- TRUE
execution_error <- NULL

tryCatch(
  {
    source(submission_path, local = env)
  },
  error = function(e) {
    execution_ok <<- FALSE
    execution_error <<- conditionMessage(e)
  }
)

if (!execution_ok) {
  record_test("code runs", FALSE, "Your code did not run successfully.")
  feedback <- c(feedback, paste("Execution error:", execution_error))
} else {
  record_test("code runs", TRUE, "Code executed without errors.")
  if (!exists("subset_df", envir = env, inherits = FALSE)) {
    record_test("subset_df exists", FALSE, "Define `subset_df` with the filtered rows.")
  } else {
    subset_df <- get("subset_df", envir = env, inherits = FALSE)
    expected <- data.frame(
      name = c("Bob", "Carol"),
      salary = c(65000, 72000)
    )
    is_df <- is.data.frame(subset_df)
    record_test("subset_df is data frame", is_df, "`subset_df` must be a data frame.")
    if (is_df) {
      has_columns <- identical(colnames(subset_df), c("name", "salary"))
      record_test("subset_df columns", has_columns, "subset_df should include only name and salary.")
      has_rows <- nrow(subset_df) == 2
      record_test("subset_df rows", has_rows, "subset_df should contain Bob and Carol.")
      matches_values <- identical(as.character(subset_df$name), as.character(expected$name)) &&
        identical(as.numeric(subset_df$salary), as.numeric(expected$salary))
      record_test("subset_df values", matches_values, "subset_df values do not match expected output.")
    }
  }
}

passed_all <- length(test_results) > 0 && all(vapply(test_results, function(t) isTRUE(t$passed), logical(1)))
if (passed_all) {
  status <- "passed"
  feedback <- c(feedback, "Correct solution submitted.")
} else if (length(feedback) == 0) {
  feedback <- c(feedback, "Filter employees with age > 30 and keep only name/salary columns.")
}

output_lines <- c(paste0("STATUS:", status))
for (msg in feedback) {
  output_lines <- c(output_lines, paste0("FEEDBACK:", msg))
}
for (t in test_results) {
  pass_label <- if (isTRUE(t$passed)) "pass" else "fail"
  output_lines <- c(output_lines, paste0("TEST:", t$name, "|", pass_label, "|", t$message))
}

writeLines(output_lines, con = output_path)

File: content/exercises/intro-python/functions/define-function/checker.py
import io
import math
import runpy
import sys


def main() -> None:
    submission_path = sys.argv[1]
    output_path = sys.argv[2]
    tests: list[tuple[str, bool, str]] = []
    feedback: list[str] = []
    status = "failed"

    try:
        namespace = runpy.run_path(submission_path, run_name="__main__")
        tests.append(("code runs", True, "Code executed without errors."))
    except Exception as error:  # noqa: BLE001
        namespace = {}
        tests.append(("code runs", False, "Code did not run successfully."))
        feedback.append(f"Execution error: {error}")

    func = namespace.get("compound_growth")
    if callable(func):
        tests.append(("function exists", True, "compound_growth function is defined."))
        try:
            actual = func(1000, 0.05, 10)
            expected = 1000 * (1 + 0.05) ** 10
            if isinstance(actual, (int, float)) and math.isclose(actual, expected, rel_tol=1e-9):
                tests.append(("function result", True, "compound_growth returns the correct value."))
            else:
                tests.append(
                    (
                        "function result",
                        False,
                        "compound_growth should return principal * (1 + rate) ** years.",
                    )
                )
        except Exception as error:  # noqa: BLE001
            tests.append(("function result", False, f"Calling compound_growth raised an error: {error}"))
    else:
        tests.append(("function exists", False, "Define a function named compound_growth."))

    passed_all = tests and all(item[1] for item in tests)
    if passed_all:
        status = "passed"
        feedback.append("Correct solution submitted.")
    elif not feedback:
        feedback.append("Define compound_growth(principal, rate, years) using the compound growth formula.")

    output_stream = io.StringIO()
    output_stream.write(f"STATUS:{status}\n")
    for item in feedback:
        output_stream.write(f"FEEDBACK:{item}\n")
    for name, passed, message in tests:
        output_stream.write(f"TEST:{name}|{'pass' if passed else 'fail'}|{message}\n")

    with open(output_path, "w", encoding="utf-8") as handle:
        handle.write(output_stream.getvalue())


if __name__ == "__main__":
    main()
